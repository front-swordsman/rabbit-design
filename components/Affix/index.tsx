import * as React from 'react'
import { AffixProps } from 'types/affix.d'
import { throttle } from '../utils/tool'

const { useState, useEffect, useRef } = React

const Affix = (userProps: AffixProps, ref: any) => {
  const props = {
    ...userProps
  }
  const { offsetTop, offsetBottom, children, target, onChange, className, style } = props

  const placeholderRef = ref || useRef(null)
  const wrapperRef = useRef(null)

  const [positionStyle, setPositionStyle] = useState({})
  // 滚动元素
  let scrollElm: Window | HTMLElement = window
  // 是否是绝对布局模式
  let fixed = false

  const handleScroll = () => {
    const rect = placeholderRef.current.getBoundingClientRect()
    let { top, bottom } = rect
    const style: React.CSSProperties = {}
    let containerTop = 0 // 容器距离视口上侧的距离
    let containerBottom = 0 // 容器距离视口下侧的距离

    if (scrollElm !== window) {
      const containerRect = (scrollElm as HTMLElement).getBoundingClientRect()
      containerTop = containerRect.top
      containerBottom = containerRect.bottom

      top = top - containerTop // 距离容器顶部的距离
      bottom = containerBottom - bottom // 距离容器底部的距离
    } else {
      bottom = window.innerHeight - bottom
    }

    if (top <= offsetTop || bottom <= offsetBottom) {
      if (!fixed) {
        style.position = 'fixed'
        style.top = offsetTop !== undefined ? offsetTop + containerTop : null
        style.bottom =
          offsetBottom !== undefined
            ? scrollElm !== window
              ? window.innerHeight - (containerBottom - offsetBottom)
              : bottom
            : null

        // 在子节点移开父节点后保持原来占位
        const { width, height } = wrapperRef.current.getBoundingClientRect()
        placeholderRef.current.style.height = `${height}px`
        placeholderRef.current.style.width = `${width}px`
        onChange && onChange(true)
        fixed = true
        setPositionStyle(style)
      }
    } else {
      if (fixed) {
        style.position = 'relative'
        onChange && onChange(false)
        fixed = false
        setPositionStyle(style)
      }
    }
  }

  const scroll = throttle(handleScroll, 20)

  useEffect(() => {
    if (target) scrollElm = target()
    // 让按钮点击调整 offsetTop 操作立马生效
    handleScroll()
    ;(scrollElm as any).addEventListener('scroll', scroll)

    return () => {
      ;(scrollElm as any).removeEventListener('scroll', scroll)
    }
  }, [offsetTop, offsetBottom])

  return (
    <div ref={placeholderRef} style={style} className={className}>
      <div ref={wrapperRef} style={positionStyle}>
        {children}
      </div>
    </div>
  )
}

export default React.forwardRef(Affix)
