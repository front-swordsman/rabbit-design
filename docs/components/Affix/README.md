---
imports:
  import AffixDemoCode from '!raw-loader!./simpleAffix.tsx';
  import AffixDemo from './simpleAffix.tsx';
---

# Affix 固钉

将页面元素钉在可视范围。当内容区域比较长，需要滚动页面时，这部分内容对应的操作或者导航需要在滚动范围内始终展现。常用于侧边菜单和按钮组合。

### 简单示例

:::demo
<Block code={AffixDemoCode} des="固钉">
  <AffixDemo />
</Block>
:::

## Props
| 参数 | 说明 | 类型 | 默认值 |
|:-|:-|:-|:-|
| offsetBottom | 距离窗口底部达到指定偏移量后触发 | number | - |
| offsetTop | 距离窗口顶部达到指定偏移量后触发 | number | - |
| onChange | 固定状态改变时触发的回调函数 | Function(affixed) | - |
| target | 设置 Affix 需要监听其滚动事件的元素 | () => HTMLElement | () => window |
