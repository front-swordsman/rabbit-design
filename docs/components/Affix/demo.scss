.scrollable-container {
  height: 100px;
  overflow-y: scroll;
}

.background1 {
  padding-top: 60px;
  height: 300px;
  background-image: url(https://zos.alipayobjects.com/rmsportal/RmjwQiJorKyobvI.jpg);
}

.background2 {
  padding-top: 120px;
  height: 300px;
  background-image: url(https://zos.alipayobjects.com/rmsportal/RmjwQiJorKyobvI.jpg);
}