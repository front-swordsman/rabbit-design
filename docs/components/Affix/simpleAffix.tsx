import * as React from 'react'
import Button from 'components/Button'
import Affix from 'components/Affix'
import './demo.scss'

const { useState } = React

function SimpleAffix() {
  const [top, setTop] = useState(100)
  const [bottom, setBottom] = useState(100)
  let container1, container2

  return (
    <>
      <h3>基础用法</h3>
      <p>最简单的用法。</p>
      <Affix
        offsetTop={top}
        style={{ marginTop: 10 }}
      >
        <Button
          type="primary"
          onClick={() => setTop((top: number) => top + 10)}
        >
          Affix Top
        </Button>
      </Affix>
      <Affix
        offsetBottom={bottom}
        style={{ marginTop: 10 }}
      >
        <Button
          type="primary"
          onClick={() => setBottom((bottom: number) => bottom + 10)}
        >
          Affix Bottom
        </Button>
      </Affix>
      <h3>固定状态改变的回调</h3>
      <p>可以获得是否固定的状态。</p>
      <Affix
        offsetTop={150}
        onChange={affixed => console.log(affixed)}
      >
        <Button
          type="primary"
        >
          150px to affix top
        </Button>
      </Affix>
      <h3>滚动容器</h3>
      <p>用 target 设置 Affix 需要监听其滚动事件的元素，默认为 window。</p>
      <div className="scrollable-container" ref={(node) => { container1 = node; }}>
        <div className="background1">
          <Affix target={() => container1} offsetTop={0}>
            <Button type="primary">
              Fixed at the top of container
            </Button>
          </Affix>
        </div>
      </div>
      <div className="scrollable-container" ref={(node) => { container2 = node; }} style={{ marginTop: 10 }}>
        <div className="background2">
          <Affix target={() => container2} offsetBottom={0}>
            <Button type="primary">
              Fixed at the Bottom of container
            </Button>
          </Affix>
        </div>
      </div>
    </>
  )
}

export default SimpleAffix