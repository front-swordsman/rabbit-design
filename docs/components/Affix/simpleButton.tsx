import * as React from 'react'
import Button from 'components/Button'
import './demo.scss'

export default class SimpleButton extends React.Component {
  render() {
    return (
      <div className="btn-demo-page">
        <h2>按钮 button</h2>
        <h3>（1）按钮类型展示</h3>
        <Button>确认</Button>
        <Button type="gray">取消</Button>
        <Button type="warn">警告</Button>
      </div>
    )
  }
}
