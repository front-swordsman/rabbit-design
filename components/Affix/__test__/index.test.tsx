import * as React from 'react'
// import { render, fireEvent } from 'react-testing-library'
import { render } from 'react-testing-library'
import Affix from '../index'

describe('render Affix', () => {
  it('render default', () => {
    const { container } = render(<Affix offsetTop={0}>Affix Top</Affix>)
    expect(container).toMatchSnapshot()
  })
  // 如何测试 getBoundingCientRect, 后续填坑。
  // it('fire scroll', () => {
  //   const { getByText } = render(<Affix style={{ marginTop: 100 }} offsetTop={0}>Affix Top</Affix>)
  //   // fireEvent.scroll(window, {
  //   //   top: 0
  //   // })

  //   expect(getByText('Affix Top').parentNode.parentNode).toHaveStyle('position: relative')
  // })
})
