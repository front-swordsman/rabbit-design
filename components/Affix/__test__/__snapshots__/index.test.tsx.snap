// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Affix render default 1`] = `
<div>
  <div
    style="height: 0px; width: 0px;"
  >
    <div
      style="position: fixed; top: 0px;"
    >
      Affix Top
    </div>
  </div>
</div>
`;
